specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,391 @@
+do so and appropriately handle the consequences of the action.
+
+If the gas paying token address is set to the `ETHER_TOKEN_ADDRESS` then the system is configured to use `ether` as the
+native asset on L2. Note that if nothing is stored in state at the `GAS_PAYING_TOKEN_SLOT` then it is considered that
+`ether` is used to pay for gas and getters on smart contracts SHOULD return the `ETHER_TOKEN_ADDRESS`.
+
+```mermaid
+---
+title: Custom Gas Token Configuration
+---
+flowchart LR
+    subgraph Layer One
+        OP[OptimismPortal]
+        subgraph SystemConfig
+          TA1[Token Address]
+        end
+        CO[Chain Operator] -->|"initialize()"| SystemConfig
+        SystemConfig -->|"setGasPayingToken()"| OP
+    end
+
+    subgraph Layer Two
+        subgraph L1Block
+          TA2[Token Address]
+        end
+    end
+    OP -->|"setGasPayingToken()"| L1Block
+```
+
+## Contract Modifications
+
+### OptimismPortal
+
+The `OptimismPortal` is updated with a new interface specifically for depositing custom tokens.
+
+#### `depositERC20Transaction`
+
+The `depositERC20Transaction` function is useful for sending custom gas tokens to L2. It is broken out into its
+own interface to maintain backwards compatibility with chains that use `ether`, to help simplify the implementation
+and to prevent footguns when trying to deposit an ERC20 native asset into a chain that uses `ether` as its native
+asset.
+
+```solidity
+function depositERC20Transaction(
+    address _to,
+    uint256 _mint,
+    uint256 _value,
+    uint64 _gasLimit,
+    bool _isCreation,
+    bytes memory _data
+) public;
+```
+
+This function MUST revert when `ether` is the L2's native asset. It MUST not be `payable`, meaning that it will
+revert when `ether` is sent with a `CALL` to it. It uses a `transferFrom` flow, so users MUST first `approve()`
+the `OptimismPortal` before they can deposit tokens.
+
+##### Function Arguments
+
+The following table describes the arguments to `depositERC20Transaction`
+
+| Name | Type | Description |
+|---------|---------|---------|
+| `_to` | `address` | The target of the deposit transaction |
+| `_mint` | `uint256` | The amount of token to deposit |
+| `_value` | `uint256` | The value of the deposit transaction, used to transfer native asset that is already on L2 from L1 |
+| `_gasLimit` | `uint64` | The gas limit of the deposit transaction |
+| `_isCreation` | `bool` | Signifies the `_data` should be used with `CREATE` |
+| `_data` | `bytes` | The calldata of the deposit transaction |
+
+#### `depositTransaction`
+
+This function is the hot code path for all usage of the `L1StandardBridge` and `L1CrossDomainMessenger`, so it MUST
+be as backwards compatible as possible. It MUST revert for chains using custom gas token when `ether` is sent with
+the `CALL`. This prevents `ether` from being stuck in the `OptimismPortal`, since it cannot be used to mint native
```

Comment:
is there a risk of funds becoming stuck if the chain transitions to a custom gas token AFTER an eth deposit but BEFORE the eth deposit is processed in the L2?

Reply:
We should think of chains starting always as custom gas token chains and not transitioning between ether and custom gas token